File: src/assets/SeatAllocation.java
package src.assets;
import java.util.*;

public class SeatAllocation {
  static boolean[] seats;
  static int vacantSeats;

  public static void main(String[] args) {
    seats = new boolean[80];
    vacantSeats = 67;

    Arrays.fill(seats, 0, 10, true);
    Arrays.fill(seats, 77, 80, true);
    display();

    String result = bookSeats(8);
    System.out.println(result);

    display();

    boolean[][] Result = transform(seats, 7);
    for (int i = 0; i < Result.length; i++) {
      for (int j = 0; j < Result[i].length; j++) {
        System.out.print(Result[i][j] + "\t");
      }
      System.out.println();
    }
  }

  static String bookSeats(int seatsToBook) {
    String seatsBooked = "";

    if (seatsToBook > vacantSeats) {
      return "Seats unavailable";
    }

    boolean availableInSameRow = false;

    // check for same row
    int rowIndex = getIndexSameRow(seatsToBook);
    if (rowIndex != -1) {
      availableInSameRow = true;
    }

    System.out.println("Available in same row: " + availableInSameRow);

    if (availableInSameRow == true) {
      // book seats
      int temp = seatsToBook;

      for (int j = rowIndex; j < rowIndex + 7 && j < 80 && temp > 0; j++) {
        if (seats[j] == false) {
          seats[j] = true;
          temp--; // seat booked
          seatsBooked += j + " ";
        }
      }
    } else {
      // check for nearest seats
      int temp = seatsToBook;

      int startingIndex = getIndexNearestSeats(seatsToBook);

      for (int j = startingIndex; temp > 0 && j < 80; j++) {
        if (seats[j] == false) {
          seats[j] = true;
          temp--; // seat booked
          seatsBooked += j + " ";
        }
      }
    }

    vacantSeats -= seatsToBook;
    printVacantSeats();

    return seatsBooked;
  }

  private static int getIndexNearestSeats(int seatsToBook) {
    int size = Integer.MAX_VALUE;
    int startIndex = -1;

    int count = 0;
    int start, end;
    start = end = 0;

    for (int i = 0; i < 80; i++) {
      end = i;

      if (seats[i] == false) {
        count++;

        // remove preceding already booked seats from window
        if (count >= seatsToBook) {
          while (seats[start] == true) {
            start++;
          }
        }

        if (count == seatsToBook) {
          int ans = end - start + 1;

          if (size > ans) {
            startIndex = start;
            size = ans;
          }
        } else if (count > seatsToBook) {
          start++;
          int ans = end - start + 1;

          if (size > ans) {
            startIndex = start;
            size = ans;
          }
        }

      }
    }
    return startIndex;
  }

  private static void printVacantSeats() {
    System.out.println("Vacant seats --> " + vacantSeats);
  }

  // returns starting index of row
  private static int getIndexSameRow(int seatsToBook) {
    int rowIndex = -1;

    for (int i = 0; i < 80; i += 7) {
      int vacant = 0; // in same row

      for (int j = i; j < i + 7 && j < 80; j++) {
        if (seats[j] == false) {
          vacant++;
        }
      }

      if (vacant >= seatsToBook) {
        rowIndex = i;
        break;
      }
    }

    return rowIndex;
  }

  private static void display() {
    for (int i = 0; i < 80; i += 7) {
      System.out.print("Row " + ((i / 7) + 1) + ": ");
      for (int j = i; j < i + 7 && j < 80; j++) {
        System.out.print(seats[j] + "\t");
      }
      System.out.println();
    }
  }

  private static boolean[][] transform(boolean[] seats, int COLS) {
    int n = seats.length;
    int ROWS = n / COLS;

    if (n % COLS > 0) {
      ROWS++;
    }

    boolean[][] result = new boolean[ROWS][COLS];

    for (int i = 0; i < n; i += COLS) {
      int ROW = i / COLS;
      result[ROW] = Arrays.copyOfRange(seats, i, i + COLS);

      if (ROW == 11) { // last row with 3 seats
        result[11] = Arrays.copyOfRange(result[11], 0, 3);
      }
    }

    return result;
  }
}

File: src/assets/SeatAllocation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.bookSeats = void 0;
var seats, vacantSeats;
function Driver() {
    seats = new Array(80);
    vacantSeats = 57;
    seats.fill(false);
    seats.fill(true, 0, 10);
    seats.fill(true, 77, 80);
    seats[1 * 7 + 5] = true;
    seats[2 * 7 + 2] = true;
    seats[3 * 7] = true;
    seats[4 * 7] = true;
    seats[5 * 7] = true;
    seats[6 * 7] = true;
    seats[7 * 7] = true;
    seats[8 * 7] = true;
    seats[9 * 7] = true;
    seats[10 * 7] = true;
    display();
    var result = bookSeats(8);
    console.log("For input 8:", result);
    result = bookSeats(3);
    console.log("For input 3:", result);
    result = bookSeats(5);
    console.log("For input 5:", result);
    result = bookSeats(7);
    console.log("For input 7:", result);
    display();
    // 2D matrix
    var Result = transform(seats, 7);
    for (var i = 0; i < Result.length; i++) {
        for (var j = 0; j < Result[i].length; j++) {
            process.stdout.write(Result[i][j] + "\t");
        }
        console.log('\n');
    }
}
function bookSeats(seatsToBook) {
    if (seatsToBook > 7 || seatsToBook <= 0) {
        return "Invalid input. Seats to book can range from 1 to 7.";
    }
    var seatsBooked = "";
    if (seatsToBook > vacantSeats) {
        return "Seats unavailable";
    }
    // check for same row
    var startingIndex = getIndexSameRow(seatsToBook);
    console.log("Available in same row:", (startingIndex != -1) ? true : false);
    if (startingIndex == -1) {
        // check for nearest seats
        startingIndex = getIndexNearestSeats(seatsToBook);
    }
    // book seats
    seatsBooked = setVacantToFilled(startingIndex, seatsToBook);
    vacantSeats -= seatsToBook;
    return seatsBooked;
}
exports.bookSeats = bookSeats;
function setVacantToFilled(startIndex, seatsToBook) {
    var seatsBooked = "";
    for (var i = startIndex; seatsToBook > 0 && i < 80; i++) {
        if (seats[i] == false) {
            seats[i] = true;
            seatsToBook--; // seat booked
            seatsBooked += i + " ";
        }
    }
    return seatsBooked;
}
function getIndexNearestSeats(seatsToBook) {
    var minSize = Infinity;
    var startIndex = -1;
    var count = 0;
    var start = 0;
    var end = 0;
    for (var i = 0; i < 80; i++) {
        end = i;
        if (seats[i] == false) {
            count++;
            if (count == seatsToBook) {
                // remove preceding already booked seats from window
                while (seats[start] == true) {
                    start++;
                }
                var size = end - start + 1;
                if (size < minSize) {
                    // got a smaller window size containing seatToBook
                    startIndex = start;
                    minSize = size;
                }
            }
            else if (count > seatsToBook) {
                // lose a vacant seat
                start++;
                count--;
                // re-check for the current i
                i--;
                count--;
            }
        }
    }
    return startIndex;
}
// returns starting index of row
function getIndexSameRow(seatsToBook) {
    var rowIndex = -1;
    for (var i = 0; i < 80; i += 7) {
        var vacant = 0; // in same row
        for (var j = i; j < i + 7 && j < 80; j++) {
            if (seats[j] == false) {
                vacant++;
            }
        }
        if (vacant >= seatsToBook) {
            rowIndex = i;
            break;
        }
    }
    return rowIndex;
}
function transform(seats, COLS) {
    var n = seats.length;
    var ROWS = n / COLS;
    if (n % COLS > 0) {
        ROWS++;
    }
    var result = [];
    for (var i = 0; i < n; i += COLS) {
        var ROW = i / COLS;
        result[ROW] = seats.slice(i, i + COLS);
    }
    return result;
}
exports.transform = transform;
function display() {
    for (var i = 0; i < 80; i += 7) {
        process.stdout.write("Row " + ((i / 7) + 1) + ": ");
        for (var j = i; j < i + 7 && j < 80; j++) {
            process.stdout.write(seats[j] + "\t");
        }
        console.log('\n');
    }
}
Driver();
